### ISSUE TITLE
`RecursionError` on simple example

### ISSUE THREAD

**Issue by rlouf** [pythological/kanren#59]
The following code fails with a `RecursionError`:

```python
from kanren import eq, lall, lany, run, vars

observations = [(0, 1), (1, 0)]

def generate(pixels):
    pairs = [(a, b) for a, b in zip(pixels[:-1], pixels[1:])]
    return lall(
        *[lany(*[eq(obs, pair) for obs in observations]) for pair in pairs]
    )

N = 1000
pixels = vars(N)
res = run(1, pixels, generate(pixels))
```

**Comment by rlouf** [comment]
It [appears](https://github.com/pytoolz/toolz/issues/188) that `toolz` does not use trampoline evaluation, and has no intention to do so. We may need to implement a version of `interleave` that uses it to avoid this `RecursionError`.

**Comment by brandonwillard** [comment]
> It [appears](https://github.com/pytoolz/toolz/issues/188) that `toolz` does not use trampoline evaluation, and has no intention to do so. We may need to implement a version of `interleave` that uses it to avoid this `RecursionError`.

Yes, that's the same type of thing I did in [`etuples`](https://github.com/pythological/etuples/blob/1ed56e5226051244aac4ef64feb9615af0d5bd28/etuples/core.py#L19) and [`unification`](https://github.com/pythological/unification/blob/94e489e6cf583f24a9af2312b2eefd8aa4949a4d/unification/core.py#L27) to get past recursion limits/errors.

**Comment by rlouf** [comment]
Ah. I tried in a fresh environment and the error still shows on my machine.

**Comment by brandonwillard** [comment]
FYI: If I set `N` to something large, the issue appears.

Anyway, yes, we should make our own version of `interleave` that doesn't use recursion.

**Comment by wrq** [comment]
I may be way out my league here, but this could not be done with zip and enumerate? I've included a small snippet to illustrate what I mean.

```
def interleave(*iters):
  res = []
  for z in zip(*iters):
    for _,x in enumerate(z):
      res.append(x)
  return res

print(interleave([1,2,3], [4,5,6], [7,8,9]))
# returns: [1, 4, 7, 2, 5, 8, 3, 6, 9]
```

If this is a complete misunderstanding, just let me know; I do not mean to interrupt or obstruct.

**Comment by brandonwillard** [comment]
> I may be way out my league here, but this could not be done with zip and enumerate? I've included a small snippet to illustrate what I mean.
> 
> ```
> def interleave(*iters):
>   res = []
>   for z in zip(*iters):
>     for _,x in enumerate(z):
>       res.append(x)
>   return res
> 
> print(interleave([1,2,3], [4,5,6], [7,8,9]))
> # returns: [1, 4, 7, 2, 5, 8, 3, 6, 9]
> ```
> 
> If this is a complete misunderstanding, just let me know; I do not mean to interrupt or obstruct.

Now that I'm coming back to this, I don't think the problem is in `interleave` itself, as much as our use of it in `lall` -> `lconj` -> `lconj_seq` -> `bind`.  The generator produced by `lconj_seq` (well, the `lconj_seq_goal` it creates) is effectively composing `N - 1` `interleave` calls.  I think _that's_ actually the thing we need to "flatten".

**Comment by wrq** [comment]
I can see that I am in fact out of my league here then, cheers and good luck!

**Comment by brandonwillard** [comment]
> I can see that I am in fact out of my league here then, cheers and good luck!

There's a bit of context here that may make things seem foreign, but we're really just talking about generators calling generators in a particular way.

The first step toward solving this issue is to remove all the unnecessary elements and obtain a more direct illustration of the issue&mdash;if possible.

Here's a first pass at that:
```python
from kanren import var, eq
from kanren.core import lconj_seq


q_lv = var()

goal_1 = lconj_seq((eq(q_lv, i) for i in range(10000)))

goal_1
# <function kanren.core.lconj_seq.<locals>.lconj_seq_goal(S)>

state_0 = {}
state_stream_1 = goal_1(state_0)

state_stream_1
# <generator object lconj_seq.<locals>.lconj_seq_goal at 0x7f6e14f580d0>

next(state_stream_1)
# RecursionError: maximum recursion depth exceeded while calling a Python object
```

Now, we can focus exclusively on the code in [`kanren.core.lconj_seq.lconj_seq_goal`](https://github.com/pythological/kanren/blob/071c626e1e252e2d8dab44dfd85196f69c70de36/kanren/core.py#L94).

**Comment by brandonwillard** [comment]
@wrq, here's a MWE demonstrating the underlying issue:
```python
from functools import reduce
from itertools import chain
from toolz import interleave


def bind(z, g):
    # Both fail for the same reason.
    # return chain.from_iterable(map(g, z))
    return interleave(map(g, z))


def goal_constructor(i):

    def g(s):
        yield (i,) + (s,)

    return g


z = reduce(bind, (goal_constructor(i) for i in range(1, 10)), iter((0,)))

list(z)
# [(9, (8, (7, (6, (5, (4, (3, (2, (1, 0)))))))))]


z = reduce(bind, (goal_constructor(i) for i in range(1, 10000)), iter((0,)))

next(z)
# RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by wrq** [comment]
Ok. First, I'm sorry that I haven't gotten to this sooner. I'm actually a part-time Walmart worker and a stay-at-home father, not really a professional programmer, so this has been on the backburner for me.

I was able to rig this up:

```
from functools import reduce
from itertools import chain
#from toolz import interleave

def interleave(*iters):
  res = []
  for z in zip(*iters):
    for _,x in enumerate(z):
      res.append(x)
  return res

def bind(z, g):
    # Both fail for the same reason.
    # return chain.from_iterable(map(g, z))
    return interleave(map(g, z))

def goal_constructor(i):
    def g(s):
        yield (i,) + (s,)
    return g

z = reduce(bind, (goal_constructor(i) for i in range(1, 10)), iter((0,)))

print(z)        # debug
print(list(z)) # checking for data model weirdness

# [(9, (8, (7, (6, (5, (4, (3, (2, (1, 0)))))))))]

z = reduce(bind, (goal_constructor(i) for i in range(1, 10000)), iter((0,)))

print(next(next(iter(z))))
```

--- OUTPUT ---
```
[<generator object goal_constructor.<locals>.g at 0x7fa23d01cdd0>]
[<generator object goal_constructor.<locals>.g at 0x7fa23d01cdd0>]
(9999, <generator object goal_constructor.<locals>.g at 0x7fa23c954eb0>)
```

This uses my little interleave hack from before with run-of-the-mill itertools, and then I noticed that z needs to be an iterator and cannot be a regular list. This is obviously a cumbersome way of thinking about the objects, because of the original Scheme domain would allow it.

I suppose the new interleave could just "return iter(res)" ? 

I'm a little embarrassed to say that I'm not even confident that this is a helpful response and that I'm missing something fundamental about all of this.

**Comment by brandonwillard** [comment]
> Ok. First, I'm sorry that I haven't gotten to this sooner. I'm actually a part-time Walmart worker and a stay-at-home father, not really a professional programmer, so this has been on the backburner for me.

No problem; I just don't want you to get the impression that this stuff is over your head, because it's not.  It's just a rather context-specific problem with some unstated constraints.

From a quick look at your approach, it seems like you have the right idea: i.e. delay evaluation further.  This will require changes to the way that `z` is handled downstream, though.  

That approach could be feasible, as long as it doesn't dramatically change the stream processing semantics and has a manageable refactoring footprint.  In other words, all the user-facing functions involved (e.g. the goal constructors, `bind`, `ldisj_seq`, `lconj_seq`, etc.) need to provide roughly the same types of objects as they currently do.

Also, the `interleave` implementation will likely need to be a generator itself, because essentially every iterator in this context can be infinite.

**Comment by brandonwillard** [comment]
I think we might need to "flatten" this example and see what can be done from there.  That would involve combining the `interleave`, `bind`, and `reduce` into a single function that takes the `goal_constructor` iterator as an argument.
